src/devices/honeywell_cm921.c

```diff
@@ -0,0 +1,449 @@
+            data = data_append(data, "setpoint", "", DATA_DOUBLE, ((msg->payload[1] << 8) | msg->payload[2]) / 100.0, NULL);
+            break;
+        }
+        case 0x1100: {
+            UNKNOWN_IF(msg->payload_length != 5 && msg->payload_length != 8);
+            data = data_append(data, "domain_id", "", DATA_INT, msg->payload[0], NULL);
+            data = data_append(data, "cycle_rate", "", DATA_DOUBLE, msg->payload[1] / 4.0, NULL);
+            data = data_append(data, "minimum_on_time", "", DATA_DOUBLE, msg->payload[2] / 4.0, NULL);
+            data = data_append(data, "minimum_off_time", "", DATA_DOUBLE, msg->payload[3] / 4.0, NULL);
+            if (msg->payload_length == 8)
+                data = data_append(data, "proportional_band_width", "", DATA_DOUBLE, (msg->payload[5] << 8 | msg->payload[6]) / 100.0, NULL);
+            break;
+        }
+        case 0x0009: {
+            UNKNOWN_IF(msg->payload_length != 3);
+            data = data_append(data, "device_number", "", DATA_INT, msg->payload[0], NULL);
+            switch (msg->payload[1]) {
+                case 0: data = data_append(data, "failsafe_mode", "", DATA_STRING, "off", NULL); break;
+                case 1: data = data_append(data, "failsafe_mode", "", DATA_STRING, "20-80", NULL); break;
+                default: data = data_append(data, "failsafe_mode", "", DATA_STRING, "unknown", NULL);
+            }
+            break;
+        }
+        case 0x3B00: {
+            UNKNOWN_IF(msg->payload_length != 2);
+            data = data_append(data, "domain_id", "", DATA_INT, msg->payload[0], NULL);
+            data = data_append(data, "state", "", DATA_DOUBLE, msg->payload[1] / 200.0 /* 0xC8 */, NULL);
+            break;
+        }
+        default: /* Unknown command */
+            UNKNOWN_IF(true);
+    }
+    return r;
+}
+
+int parse_msg(bitbuffer_t *bmsg, int row, message_t *msg) {
+    unsigned num_bytes = bmsg->bits_per_row[0]/8;
+    unsigned ipos = 0;
+    const uint8_t *bb = bmsg->bb[row];
+    memset(msg, 0, sizeof(message_t));
+
+    // Checksum: All bytes add up to 0.
+    uint8_t bsum = 0;
+    for (unsigned i = 0; i < num_bytes; i++)
+        bsum += bitrow_get_byte(bmsg->bb[row], i*8);
+    uint8_t checksum_ok = bsum == 0;
+    msg->crc = bitrow_get_byte(bmsg->bb[row], bmsg->bits_per_row[row] - 8);
+
+    if (!checksum_ok)
+        return 0;
+
+    msg->header = next(bb, &ipos, num_bytes);
+
+    msg->num_device_ids = (msg->header >> 2) & 0x03; // total speculation.
+
+    for (unsigned i = 0; i < msg->num_device_ids; i++)
+        for (unsigned j = 0; j < 3; j++)
+            msg->device_id[i][j] = next(bb, &ipos, num_bytes);
+
+    msg->command = (next(bb, &ipos, num_bytes) << 8) | next(bb, &ipos, num_bytes);
+    msg->payload_length = next(bb, &ipos, num_bytes);
+
+    if (msg->payload_length > 0) {
+        msg->payload = (uint8_t*)malloc(sizeof(uint8_t) * msg->payload_length);
```

Comment:
Can we get rid of all allocations?

The payload is a sequence of (8-bit) bytes represented in hexadecimal (i.e. 00 to FF) and with a maximum length of 48 (minimum length 1).

Here it should be possible to replace it with a 48 bytes long array.

Reply:
Not sure 48 is the right number, but I did get rid of the allocations.